lsp-types/src/Language/LSP/VFS.hs

```diff
@@ -291,26 +291,18 @@ applyChanges = foldl' applyChange
 applyChange :: Rope -> J.TextDocumentContentChangeEvent -> Rope
 applyChange _ (J.TextDocumentContentChangeEvent Nothing Nothing str)
   = Rope.fromText str
-applyChange str (J.TextDocumentContentChangeEvent (Just (J.Range (J.Position sl sc) _to)) (Just len) txt)
-  = changeChars str start (fromIntegral len) txt
-  where
-    start = Rope.rowColumnCodeUnits (Rope.RowColumn (fromIntegral sl) (fromIntegral sc)) str
-applyChange str (J.TextDocumentContentChangeEvent (Just (J.Range (J.Position sl sc) (J.Position el ec))) Nothing txt)
-  = changeChars str start len txt
-  where
-    start = Rope.rowColumnCodeUnits (Rope.RowColumn (fromIntegral sl) (fromIntegral sc)) str
-    end = Rope.rowColumnCodeUnits (Rope.RowColumn (fromIntegral el) (fromIntegral ec)) str
-    len = end - start
+applyChange str (J.TextDocumentContentChangeEvent (Just (J.Range (J.Position sl sc) (J.Position fl fc))) _ txt)
+  = changeChars str (Rope.Position (fromIntegral sl) (fromIntegral sc)) (Rope.Position (fromIntegral fl) (fromIntegral fc)) txt
 applyChange str (J.TextDocumentContentChangeEvent Nothing (Just _) _txt)
   = str
 
 -- ---------------------------------------------------------------------
 
-changeChars :: Rope -> Int -> Int -> Text -> Rope
-changeChars str start len new = mconcat [before, Rope.fromText new, after']
+changeChars :: Rope -> Rope.Position -> Rope.Position -> Text -> Rope
+changeChars str start finish new = mconcat [before', Rope.fromText new, after]
   where
-    (before, after) = Rope.splitAt start str
-    after' = Rope.drop len after
+    (before, after) = fromJust $ Rope.splitAtPosition finish str
+    (before', _) = fromJust $ Rope.splitAtPosition start before
```

Comment:
What's the best way to handle malformed input here? Since LSP counts by UTF-16 code units (not code points), it is possible that split happens in the middle of a code point, in which case `text-lines` returns `Nothing`. 
(`rope-utf16-splay` in such situation silently extends split region until the end of a code point)

Reply:
@alanz ? do you remember what's up here?

Reply:
Nope. To be honest I have never really paid attention to issues of splits in the middle of code points, assuming the client will send meaningful input, since they are managing a cursor over the text and sending its positions.

Reply:
We *could* throw. Maybe blowing up with an informative error message is the most sensible thing to do here.

Reply:
I would be happy with either:
- Throwing, based on judging this to be very unlikely.
- Returning `Nothing`, and propagating that up to the exported functions. In `changeFromClientVFS` we can drop the change and log, as is done in the existing `Nothing` branch.

Thoughts @alanz ?

Reply:
Alternatively, we can merge this with the `fromJust` and experiment with removing it with @Bodigrim out of the loop :)

Reply:
Looking more at the code, I think we could push these failures right the way up to the LSP message response, which is probably the right thing to do, but I think we should do it separately. So my vote goes for "leave `fromJust` for now and remove subsequently", which I volunteer to look into.

Reply:
I think this is unlikely to happen in practice, so leaving as is would likely be safe, as is pushing it up to an LSP response. For the latter it at least makes us aware it is happening, and hopefully that will show up in initial developer dogfooding, or not at all.